tests/e2e/utils/request-mock.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * Observer to listen and mock puppeteer page requests.
+ */
+export class RequestMock {
+	/**
+	 * Map of mock functions to URL patters they should apply to.
+	 *
+	 * @type {Map<jest.Mock, RegExp>}
+	 */
+	#mocks;
+	/**
+	 * The instance of a page being observed.
+	 *
+	 * @type {import("puppeteer").Page}
+	 */
+	#observedPage;
+
+	constructor() {
+		this.#mocks = new Map();
+		this.#callMocksOnThis = this.#callMocks.bind( this );
+	}
+
+	/**
+	 * Get the mock to stub/mock/observe requests for a URL matching the given `pattern`.
+	 *
+	 * @param {RegExp} pattern Pattern for an URL(s) to mock.
+	 * @return {jest.Mock<import("puppeteer").Request>} New jest mock, will be called with an intercepted request. If the mock returns:
+	 * 	- `undefined` - the original request will be passed through, unless already handled.
+	 *  - falsy - the request will be aborted
+	 *  - truthy - no action will be taken
+	 */
+	mock( pattern ) {
+		const mock = jest.fn().mockName( 'mocked request' );
+		this.#mocks.set( mock, pattern );
+		return mock;
+	}
+
+	/**
+	 * Stop applying all or the given mock.
+	 *
+	 * @param {jest.Mock} [mock] Mock to be restored/unobserved. If not given all will be restored.
+	 */
+	restore( mock ) {
+		if ( mock ) {
+			this.#mocks.delete( mock );
+		} else {
+			this.#mocks.clear();
+		}
+	}
+
+	/**
+	 * Start oberving/intercepting requests for the given `page`.
+	 *
+	 * @param {import("puppeteer").Page} page
+	 */
+	async observe( page ) {
+		// Unobserve the previous page.
+		this.disconnect();
+
+		this.#observedPage = page;
+		await page.setRequestInterception( true );
+		page.on( 'request', this.#callMocksOnThis );
+	}
+	/**
+	 * Stop observing.
+	 */
+	disconnect() {
+		if ( this.#observedPage ) {
+			this.#observedPage.off( 'request', this.#callMocksOnThis );
+			this.#observedPage = undefined;
+		}
+	}
+
+	/**
+	 * Call a matching mock, if any.
+	 *
+	 * @param {import("puppeteer").Request} interceptedRequest
+	 */
+	#callMocks( interceptedRequest ) {
```

Comment:
❔ 
I tested e2e test by replacing the line with `#callMocksOnThis = ( interceptedRequest ) => {`, and it works. Is there any concern about not adopting this syntax?

Reply:
yes, as then the same function will be shared between all instances when it's added/removed by the event listener. So If one instance will remove its listener, it will make all other instances stop listening.

Reply:
If the puppeteer API would support [`EventListener`s](https://developer.mozilla.org/en-US/docs/Web/API/EventListener) in `page.addEventListener`, we could use it like:
```js
	async observe( page ) {
		// Unobserve the previous page.
		this.disconnect();

		this.#observedPage = page;
		await page.setRequestInterception( true );
		page.addEventListener( 'request', this );
	}
	disconnect() {
		if ( this.#observedPage ) {
			this.#observedPage.removeEventListener( 'request', this );
			this.#observedPage = undefined;
		}
	}

	/**
	 * Call a matching mock, if any.
	 *
	 * @param {import("puppeteer").Request} interceptedRequest
	 */
	#handleEvent( interceptedRequest ) {
```
without the need for binding an instance of a function to the `RequestMock` instance.

Like for any other Web listeners https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

Reply:
Sorry for not being clear, what I meant was the method syntax `method = () => {}` is still binding `this` on the function itself, so it would not be a prototype method. In this way, it could make it no need to bind the prototype method `callMocks` in the constructor and an additional member `callMocksOnThis`. I think it should be equivalent to the current implementation.

For example:

```js
class Foo {
	msg = 'Hi';
	bar = () => console.log( this.msg );
}
```

It would not be shared between all instances and could be invoked as a normal function call directly.

![image](https://user-images.githubusercontent.com/17420811/130557225-7baee06b-5f42-45b2-b29f-3feffad28f0c.png)

Reply:
ah, ok. So you mean remove it from class/prototype and add to the instance only?

Sure it will work but will be slightly less extendable, as we will have no function to inherit by things like `WCSMock`, `GoogleAPIMock`, etc. that was thinking we could have one day.
But as they are not here today, and I still struggle with bigger problems to implement them, maybe we can simplify the current shape for the sake of YAGNI, or TheyAintGonnaByUsefull ;)

Reply:
> Sure it will work but will be slightly less extendable, as we will have no function to inherit by things like `WCSMock`, `GoogleAPIMock`, etc. that was thinking we could have one day.

Since I rarely use JavaScript inheritance in my practical experience except for exercise, I don't have a concrete opinion about that. I was just curious why you chose to use the current implementation. I'm good with the current way.

Reply:
I changed it in [`dabc336`](https://github.com/woocommerce/google-listings-and-ads/pull/970/commits/dabc33650c2482a0d2bd0ab6cc7f14cf551d4966) to 
```js
#callMock;
constructor() {
	this.#mocks = new Map();
	/**
	 * Calls a matching mock, if any.
	 *
	 * @param {import("puppeteer").Request} interceptedRequest
	 */
	this.#callMock = ( interceptedRequest ) => {
```
Unfortunately, VSCode's JSDoc fails to pick up private instance methods, leaving absolutely no hint for `#callMock`property which is assigned to a function statement.

We can change it to `_callMock`, to get IDE support, as private methods are not well addopted anyway.